plutus-tx/compiler/Language/PlutusTx/Data.hs

```diff
@@ -0,0 +1,55 @@
+{-# LANGUAGE LambdaCase #-}
+module Language.PlutusTx.Data where
+
+import           Prelude              hiding (fail)
+
+import           Data.Bifunctor       (bimap)
+import           Data.Bitraversable   (bitraverse)
+import           Data.ByteString.Lazy as BSL
+
+import           Control.Monad.Fail
+
+import qualified Codec.CBOR.Term      as CBOR
+import qualified Codec.Serialise      as Serialise
+
+
+-- | A generic "data" type.
+--
+-- The main constructor 'Constr' represents a datatype value in sum-of-products
+-- form: @Constr i args@ represents a use of the @i@th constructor along with its arguments.
+--
+-- The other constructors are various primitives.
+data Data =
+      Constr Integer [Data]
+    | Map [(Data, Data)]
+    | I Integer
+    | B ByteString
```

Comment:
I've already forgotten that one can hardcode primitive types instead of being generic over universes :)